### PR TITLE
Add Hazelcast online store implementation [API-1778]

### PR DESCRIPTION
**Pull request by mehmettokgoz**
The implementation is under `sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py`.

### CODE REVIEW

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+class HazelcastOnlineStore(OnlineStore):
+    _client: Optional[HazelcastClient] = None
+
+    def _get_client(self, config: HazelcastOnlineStoreConfig):
+        logger = logging.getLogger("hazelcast")
+        logger.setLevel(logging.ERROR)
+        if not self._client:
+            if config.cloud:
+                self._client = HazelcastClient(
```
Comment (mehmettokgoz) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:63]: Since we don't have public `Config()` implementation yet, I had to create cloud enabled and ssl enabled clients separately.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        for entity_key, features, timestamp, created_ts in data:
+            entity_key_bin = serialize_entity_key(
+                entity_key,
+                entity_key_serialization_version=2,
+            ).hex()
+            timestamp = _convert_to_naive_utc(timestamp)
+            if created_ts is not None:
+                created_ts = _convert_to_naive_utc(created_ts)
+            for feature_name, feature_value in features.items():
+                client.sql.execute(query,
```
Comment (mehmettokgoz) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:122]: Put operation is done via SQL query since we are using JSON format as value. I don't know if simple `map.put()` operation is faster. Any comments?

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                                   entity_key_bin,
+                                   feature_name,
+                                   feature_value.SerializeToString().hex(),
```
Comment (mehmettokgoz) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:125]: `feature_value.SerializeToString()` returns bytes object. AFAIK, there is no way to store bytes in Hazelcast cluster, so I have converted it to hex.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        requested_features: Optional[List[str]] = None
+    ) -> List[Tuple[Optional[datetime], Optional[Dict[str, ValueProto]]]]:
+
+        online_store_config = config.online_store
+        assert isinstance(online_store_config, HazelcastOnlineStoreConfig)
+
+        client = self._get_client(online_store_config)
+
+        result: List[Tuple[Optional[datetime], Optional[Dict[str, Any]]]] = []
+        requested_features_str = ", ".join(f"'{f}'" for f in requested_features)
```
Comment (mehmettokgoz) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: We have to add this `requested_features` check either in here or in line 160 since user might want to read only some of the features, not all of them. Notice that if we check this at line 160 (after retrieving the features), there is a cost of retrieving unnecessary feature.

Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: The format conversion between `requested_features` to `requested_features_str` is below:

```
requested_features = ["acc_percentage", "conv_percentage"]
requested_features_str = "('acc_percentage', 'conv_percentage')"
```
This makes possible to use `IN` statement in SQL query.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        query = f"""
+                SELECT feature_name, feature_value, event_ts 
+                FROM {_feature_map_name(config.project, table)} 
+                WHERE entity_key=? AND feature_name IN ({requested_features_str})"""
+        for entity_key in entity_keys:
+            entity_key_bin = serialize_entity_key(
+                entity_key,
+                entity_key_serialization_version=2,
+            ).hex()
```
Comment (mehmettokgoz) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Same problem about `hex()` with https://github.com/mehmettokgoz/feast/pull/1/files#r1055201988.

File: docs/reference/online-stores/hazelcast.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+* This implementation inherits all strengths of Hazelcast such as high availability, fault-tolerance, and data distribution.
+* Secure TSL/SSL connection is supported by Hazelcast online store.
+* You can set TTL (Time-To-Live) setting for your features in Hazelcast cluster. 
+
+Each feature view corresponds to an IMap in Hazelcast cluster and the entries in that IMap corresponds to features of entities.
+Each feature value stored separately and can be retrieved individually.
+
+## Getting started
+
+In order to use Hazelcast online store, you'll need to run `pip install 'feast[hazelcast]'`. You can then get started with the command `feast init REPO_NAME -t hazelcast`.
```
Comment (yuce) [inline docs/reference/online-stores/hazelcast.md:20]: It would be nice to mention that the user needs to have a running Hazelcast cluster and add a link to: https://hazelcast.com/get-started/

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+from feast.protos.feast.types.Value_pb2 import Value as ValueProto
+from feast.repo_config import FeastConfigBaseModel
+from feast.usage import log_exceptions_and_usage
+
+
+class HazelcastOnlineStoreConfig(FeastConfigBaseModel):
+    """Online store config for Hazelcast store"""
+
+    type: Literal["hazelcast"] = "hazelcast"
+    """Online store type selector"""
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:60]: The documentation for instance variables should be in the class documentation. e.g., https://github.com/feast-dev/feast/blob/f2cbf43d4be6829ce3affb72b8a5416d8e084ba9/sdk/python/feast/base_feature_view.py#L29

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+    ssl_password: Optional[StrictStr] = ""
+    """Password for decrypting the keyfile if it is encrypted."""
+
+    key_ttl_seconds: Optional[int] = 0
+    """Hazelcast key bin TTL (in seconds) for expiring entities"""
+
+
+class HazelcastOnlineStore(OnlineStore):
+    _client: Optional[HazelcastClient] = None
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:96]: Is documentation needed for this class?

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+
+    def _get_client(self, config: HazelcastOnlineStoreConfig):
+        logging.getLogger("hazelcast").setLevel(logging.ERROR)
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: We shouldn't set the logger parameters, the user is supposed to do that.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        if not self._client:
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: We should protect `_client` with a mutex, otherwise more than one client can be created if `_get_client` is called when the client is connecting to the cluster.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+        self,
+        config: RepoConfig,
+        table: FeatureView,
+        data: List[
+            Tuple[EntityKeyProto, Dict[str, ValueProto], datetime, Optional[datetime]]
+        ],
+        progress: Optional[Callable[[int], Any]],
+    ) -> None:
+        online_store_config = config.online_store
+        assert isinstance(online_store_config, HazelcastOnlineStoreConfig)
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: What's the expected exception if `online_store_config` is not `HazelcastOnlineStoreConfig` ? I think it's better to use `if` and raise that exception instead of relying on assert. Same comment for assert usage.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+
+        client = self._get_client(online_store_config)
+        fv_map = client.get_map(_table_id(config.project, table))
+
+        for entity_key, values, event_ts, created_ts in data:
+            entity_key_bin = serialize_entity_key(
+                entity_key,
+                entity_key_serialization_version=config.entity_key_serialization_version,
+            ).hex()
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Using base64 encode instead of hex should create shorter strings. Same comment for other hex usage.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+            event_ts_utc = _to_utc_timestamp(event_ts)
+            created_ts_utc = 0.0
+            if created_ts is not None:
+                created_ts_utc = _to_utc_timestamp(created_ts)
+            for feature_name, value in values.items():
+                feature_value = value.SerializeToString().hex()
+                __key = str(entity_key_bin) + feature_name
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Why `key` was prefixed with two underscores?

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+class HazelcastOnlineStoreConfig(FeastConfigBaseModel):
+    """Online store config for Hazelcast store"""
+
+    type: Literal["hazelcast"] = "hazelcast"
+    """Online store type selector"""
+
+    cluster_name: StrictStr = "dev"
+    """Name of the cluster you want to connect. The default cluster name is `dev`"""
+
+    cluster_members: Optional[List] = ["localhost:5701"]
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: The type should be `Optional[List[str]]` ?

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 10 above)
Comment (yuce) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: It's better to use `is` with this check:
```
if self._client is None:
```

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+        for entity_key in entity_keys:
+            feature_keys = []
+            entity_key_str = str(
+                serialize_entity_key(
+                    entity_key,
+                    entity_key_serialization_version=config.entity_key_serialization_version,
+                ).hex()
+            )
+            if requested_features:
+                for feature in requested_features:
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: You can use a list comprehension here, which is usually faster and more compact. That also allows you to get rid of `feature_keys = []` line:
```
feature_keys = [entity_key_str + feature for feature in requested_features]
```
The same can be used for the `else` branch.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+                    feature_keys.append(entity_key_str + feature)
+            else:
+                for f in table.features:
+                    feature_keys.append(entity_key_str + f.name)
+            hz_keys.extend(feature_keys)
+            entity_keys_str[entity_key_str] = feature_keys
+
+        data = fv_map.get_all(hz_keys).result()
+        entities = []
+        for _key in hz_keys:
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: We should get rid of the underscore before `key`.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+            try:
+                data[_key] = data[_key].loads()
+                entities.append(data[_key]["entity_key"])
+            except KeyError:
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:234]: Is it possible to get `KeyError`?

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+                continue
+
+        for curr_key in entity_keys_str:
+            if curr_key in entities:
+                entry = dict()
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: The Pythonic way of defining an empty dict is `{}`.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+                event_ts: Optional[datetime] = None
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Does defining the type for these local variables add any value? I think that makes the code harder to read.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+                for _key in entity_keys_str[curr_key]:
+                    row = data[_key]
+                    value = ValueProto()
+                    value.ParseFromString(bytes.fromhex(row["feature_value"]))
+                    entry[row["feature_name"]] = value
+                    event_ts = datetime.utcfromtimestamp(row["event_ts"])
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Instead of `utcfromtimestamp` the following should be used:
```
datetime.fromtimestamp(row["event_ts"], tz=timezone.utc)
```
See: https://blog.ganssle.io/articles/2019/11/utcnow.html

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+                entries.append((event_ts, entry))
+            else:
+                entries.append((None, None))
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:249]: Why is this necessary?

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+
+        for curr_key in entity_keys_str:
+            if curr_key in entities:
+                entry = dict()
+                event_ts: Optional[datetime] = None
+                for _key in entity_keys_str[curr_key]:
+                    row = data[_key]
+                    value = ValueProto()
+                    value.ParseFromString(bytes.fromhex(row["feature_value"]))
+                    entry[row["feature_name"]] = value
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: I think we need to have "constants" instead of strings for these dictionary keys.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+
+    def update(
+        self,
+        config: RepoConfig,
+        tables_to_delete: Sequence[FeatureView],
+        tables_to_keep: Sequence[FeatureView],
+        entities_to_delete: Sequence[Entity],
+        entities_to_keep: Sequence[Entity],
+        partial: bool,
+    ):
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:260]: You may want to add the return type. Same comment for other methods.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+        online_store_config = config.online_store
+        assert isinstance(online_store_config, HazelcastOnlineStoreConfig)
+
+        client = self._get_client(online_store_config)
+        project = config.project
+
+        for table in tables_to_keep:
+            client.sql.execute(
+                f"""CREATE OR REPLACE MAPPING {_table_id(project, table)} (
```
Comment (yuce) [inline sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Do we still need to create an SQL mapping? I thought you were using the map directly. Same comment for other SQL code.

**Review by yuce — CHANGES_REQUESTED** [review]
Here are some initial set of ideas...

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 7 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:60]: **All** other online store implementation follows this convention, see Redis: https://github.com/feast-dev/feast/blob/master/sdk/python/feast/infra/online_stores/redis.py#L59

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 17 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:234]: Yes possible, if user would wants to retrieve features of a non-materialized entity, then the `data` does not contains it.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 21 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:249]: Same case with `KeyError` above, the features might non-exist in the online store. In this case, `(None, None)` should added to return list.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 23 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:260]: The `update` and `teardown` does not return anything. I think it is not necessary to add `None` return type for those.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 24 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: This is for creating the map schemas of feature views in the online store, so users can see it. We can pass this step since Hazelcast allow us but I wanted to provide some functionality in here since we have to provide this function.

Also it's easier to remove all entries from Hazelcast using SQL. Functions with SQL queries won't executed very frequently so I think they don't add any overhead.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 8 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:96]: Yes, added now https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-47764bf85c9631f63e542e314d1d29579a85e570b2813040caaade67743322d7R87

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 9 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Right, removed from here but added to test class to get a cleaner test output, see: https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-82e2d2b8b2b436ed48802ec682bdddf2e057cf9b39a148528dc14580db2f58e1R20

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 10 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: I understand that but none of other OS implementations does not provide this, I thought they are ignoring this because it's a rare condition. 

Should we care and implement Singleton pattern here? Maybe I can ask to Feast community about why they don't request this in OS implementations.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 11 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: I have added a new class for HazelcastErrors and changed check to if logic: https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-47764bf85c9631f63e542e314d1d29579a85e570b2813040caaade67743322d7R253

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 13 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: I've renamed with more meaningful ones: https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-47764bf85c9631f63e542e314d1d29579a85e570b2813040caaade67743322d7R166

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 14 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Fixed at https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-47764bf85c9631f63e542e314d1d29579a85e570b2813040caaade67743322d7R64

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 15 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Makes sense, refactored at https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-47764bf85c9631f63e542e314d1d29579a85e570b2813040caaade67743322d7R215

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 16 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Fixed at https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-47764bf85c9631f63e542e314d1d29579a85e570b2813040caaade67743322d7R228.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 17 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py:234]: Actually one of the tests were failing because of this and I had to add this check.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 18 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Fixed at https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-47764bf85c9631f63e542e314d1d29579a85e570b2813040caaade67743322d7R230

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 19 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Removing this type definition since it's obvious that it's timestamp: https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-47764bf85c9631f63e542e314d1d29579a85e570b2813040caaade67743322d7R231

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 20 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Thanks for this info, fixed now: https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-47764bf85c9631f63e542e314d1d29579a85e570b2813040caaade67743322d7R237.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 22 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Makes since, this way users can change the Hazelcast schema for the entries: https://github.com/mehmettokgoz/feast/pull/1/commits/9f8174c72faf2bbdbd6f7fe4b414d63f20338600#diff-47764bf85c9631f63e542e314d1d29579a85e570b2813040caaade67743322d7R42

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 12 above)
Comment (mehmettokgoz) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: Makes sense, combined unique key is becoming too long with HEX version, replaced with base64 now.

File: docs/reference/online-stores/hazelcast.md (hunk 6 above)
Comment (mehmettokgoz) [reply docs/reference/online-stores/hazelcast.md:20]: Added here https://github.com/mehmettokgoz/feast/pull/1/commits/4cffde293b3d80ec4a4fe552c0343dbba2247d4d. Notice that also `README.md` file is added inside the `online_store` module.

File: sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py (hunk 10 above)
Comment (yuce) [reply sdk/python/feast/infra/online_stores/contrib/hazelcast_online_store/hazelcast_online_store.py]: I am not sure about why others don't add that, but if `_get_client` can be called by multiple threads, the problem I mentioned will happen. Please feel free to ask about it to the community.

Using the singleton would be OK.

**Review by yuce — APPROVED** [review]
